x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportUnfollowAction.java

```diff
@@ -178,10 +179,16 @@ private void removeRetentionLeaseForShard(
             ) {
                 logger.trace("{} removing retention lease [{}] while unfollowing leader index", followerShardId, retentionLeaseId);
                 final ThreadContext threadContext = threadPool.getThreadContext();
+                // We're about to stash the thread context for this retention lease removal. The listener will be completed while the
+                // context is stashed. The context needs to be restored in the listener when it is completing or else it is simply wiped.
+                final ActionListener<ActionResponse.Empty> preservedListener = new ContextPreservingActionListener<>(
+                    threadContext.newRestorableContext(true),
+                    listener
+                );
                 try (ThreadContext.StoredContext ignore = threadPool.getThreadContext().stashContext()) {
                     // we have to execute under the system context so that if security is enabled the removal is authorized
                     threadContext.markAsSystemContext();
-                    CcrRetentionLeases.asyncRemoveRetentionLease(leaderShardId, retentionLeaseId, remoteClient, listener);
+                    CcrRetentionLeases.asyncRemoveRetentionLease(leaderShardId, retentionLeaseId, remoteClient, preservedListener);
```

Comment:
This one is suspicious to me: this doesn't seem like a place where the context might get lost since we're just executing an action on the remote client. Again I'd rather fix this lower down the call stack to make sure it's not causing any other bugs. Sorry to say that I don't see the obvious location where the context is being lost here tho.

Reply:
The context is being lost on line 188 here - We're stashing the context before running on the remote client. When the client returns it executes the listener with the empty context, when instead it should restore the context once the retention lease removal is completed. Agreed that it would be far nicer to do the fix lower down, but that would still need the stored context to be moved down there as well.

Reply:
Oh bah I was chasing the problem all the way down the call stack and totally missed the obvious thing there. Sorry for the noise.